pkg/hive/hive.go

```diff
@@ -218,3 +191,58 @@ func (s *Service) disconnect(peer p2p.Peer) error {
 
 	return nil
 }
+
+const checkPeersTimeout = time.Minute * 15
+
+func (s *Service) checkAndAddPeers(peers pb.Peers) {
+	ctx, cancel := context.WithTimeout(context.Background(), checkPeersTimeout)
+	defer cancel()
+
+	sem := make(chan struct{}, 5)
+	var peersToAdd []swarm.Address
+
+	for i := range peers.Peers {
+		sem <- struct{}{}
+
+		go func(idx int) {
+			defer func() {
+				<-sem
+			}()
+
+			newPeer := peers.Peers[idx]
+
+			multiUnderlay, err := ma.NewMultiaddrBytes(newPeer.Underlay)
+			if err != nil {
+				s.logger.Errorf("hive: multi address underlay err: %v", err)
+				return
+			}
+
+			// check if the underlay is usable by doing a raw ping using libp2p
+			_, err = s.streamer.Ping(ctx, multiUnderlay)
+			if err != nil {
+				s.logger.Errorf("hive: multi address underlay %s not reachable err: %w", multiUnderlay, err)
+				return
+			}
+
+			bzzAddress := bzz.Address{
+				Overlay:     swarm.NewAddress(newPeer.Overlay),
+				Underlay:    multiUnderlay,
+				Signature:   newPeer.Signature,
+				Transaction: newPeer.Transaction,
+			}
+
+			err = s.addressBook.Put(bzzAddress.Overlay, bzzAddress)
+			if err != nil {
+				s.logger.Warningf("skipping peer in response %s: %v", newPeer.String(), err)
+				return
+			}
+
+			peersToAdd = append(peersToAdd, bzzAddress.Overlay)
+		}(i)
+	}
+
+	if s.addPeersHandler != nil {
```

Comment:
I think we could do this as a defer. So for some reason if the above attempts timeout after some are done, we would still add some addresses to the book.